File: Medium/Java/SortAndSearch/MergeIntervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> res = new ArrayList<>();
        intervals.sort((i1, i2) -> Integer.compare(i1.start, i2.start));
        for (int i = 0; i < intervals.size(); i ++) {
            Interval pre = intervals.get(i);
            if (res.isEmpty())
                res.add(new Interval(pre.start, pre.end));
            else {
                Interval last = res.get(res.size() - 1);
                if (pre.start <= last.end) {
                    last.start = Math.min(last.start, pre.start);
                    last.end = Math.max(last.end, pre.end);
                }
                else
                    res.add(new Interval(pre.start, pre.end));
            }
        }
        return res;
    }
}

File: Medium/Python/SortingAndSearching/MergeIntervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
            :type intervals: List[Interval]
            :rtype: List[Interval]
            """
        if len(intervals) == 0:
            return []
        intervals = sorted(intervals, lambda x, y: cmp(x.start, y.start))
        res = [intervals[0]]
        for interval in intervals[1:]:
            if interval.start <= res[-1].start:
                res[-1].start = interval.start
                res[-1].end = max(interval.end, res[-1].end)
            elif interval.start <= res[-1].end:
                res[-1].end = max(interval.end, res[-1].end)
            else:
                res.append(interval)
        return res
